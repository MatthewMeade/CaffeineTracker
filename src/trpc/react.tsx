'use client';

import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { loggerLink, unstable_httpBatchStreamLink } from '@trpc/client';
import { createTRPCReact } from '@trpc/react-query';
import { useState } from 'react';
import superjson from 'superjson';

import type { AppRouter } from '~/server/trpc/router';

export const api = createTRPCReact<AppRouter>();

function getBaseUrl() {
    if (typeof window !== 'undefined') return ''; // browser should use relative url
    if (process.env.VERCEL_URL) return `https://${process.env.VERCEL_URL}`; // SSR should use vercel url
    return `http://localhost:${process.env.PORT ?? 3000}`; // dev SSR should use localhost
}

export function TRPCReactProvider(props: {
    children: React.ReactNode;
    headers?: Headers;
}) {
    const [queryClient] = useState(() => new QueryClient());
    const [trpcClient] = useState(() =>
        api.createClient({
            links: [
                loggerLink({
                    enabled: (opts) =>
                        process.env.NODE_ENV === 'development' ||
                        (opts.direction === 'down' && opts.result instanceof Error),
                }),
                unstable_httpBatchStreamLink({
                    transformer: superjson,
                    url: `${getBaseUrl()}/api/trpc`,
                    headers() {
                        const heads = new Map(props.headers);
                        heads.set('x-trpc-source', 'react');
                        return Object.fromEntries(heads);
                    },
                }),
            ],
        }),
    );

    return (
        <api.Provider client={trpcClient} queryClient={queryClient}>
            <QueryClientProvider client={queryClient}>
                {props.children}
            </QueryClientProvider>
        </api.Provider>
    );
} 