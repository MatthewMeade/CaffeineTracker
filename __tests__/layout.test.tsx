import { describe, it, expect, vi } from "vitest";
import { render } from "@testing-library/react";
import RootLayout from "../src/app/layout";
import type { Session } from "next-auth";

// Mock next/font/google
vi.mock("next/font/google", () => ({
  Inter: () => ({
    className: "font-sans",
    variable: "--font-sans",
  }),
}));

// Use a let variable for the mock function to avoid hoisting issues
let mockAuth: ReturnType<typeof vi.fn>;
vi.mock("~/auth", () => ({
  get auth() {
    return mockAuth;
  },
}));

// Mock tRPC provider
vi.mock("~/trpc/react", () => ({
  TRPCReactProvider: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
}));

describe("RootLayout", () => {
  beforeEach(() => {
    mockAuth = vi.fn();
  });

  it("renders the layout with session provider and children", async () => {
    const mockSession: Session = {
      user: {
        id: "test-user-id",
        email: "test@example.com",
      },
      expires: "2099-01-01T00:00:00.000Z",
    };
    mockAuth.mockResolvedValueOnce(mockSession);

    const element = await RootLayout({ children: <div data-testid="layout-test-child">Test Child</div> });
    
    // Test the structure without rendering to avoid hydration issues
    expect(element).toBeDefined();
    expect(element.type).toBe('html');
    
    // Verify the children are present in the JSX structure
    const childrenElement = element.props.children;
    expect(childrenElement).toBeDefined();
    expect(childrenElement.type).toBe('body');
    
    // Find the test child in the body
    const bodyChildren = childrenElement.props.children;
    expect(bodyChildren).toBeDefined();
    
    // Snapshot test for the returned JSX structure
    expect(element).toMatchSnapshot();
  });

  it("renders the layout with null session", async () => {
    mockAuth.mockResolvedValueOnce(null);
    const element = await RootLayout({ children: <div data-testid="layout-test-child">Test Child</div> });
    
    // Test the structure without rendering to avoid hydration issues
    expect(element).toBeDefined();
    expect(element.type).toBe('html');
    
    // Verify the children are present in the JSX structure
    const childrenElement = element.props.children;
    expect(childrenElement).toBeDefined();
    expect(childrenElement.type).toBe('body');
    
    expect(element).toMatchSnapshot();
  });

  it("includes the correct metadata", () => {
    const metadata = {
      title: "Caffeine Tracker",
      description: "Track your caffeine intake",
      icons: [{ rel: "icon", url: "/favicon.ico" }],
    };
    expect(metadata).toEqual({
      title: "Caffeine Tracker",
      description: "Track your caffeine intake",
      icons: [{ rel: "icon", url: "/favicon.ico" }],
    });
  });
});
