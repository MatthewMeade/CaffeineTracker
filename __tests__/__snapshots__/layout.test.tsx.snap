// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RootLayout > renders the layout with null session 1`] = `
{
  "$$typeof": Symbol(react.transitional.element),
  "_owner": null,
  "_store": {},
  "key": null,
  "props": {
    "children": {
      "$$typeof": Symbol(react.transitional.element),
      "_owner": null,
      "_store": {},
      "key": null,
      "props": {
        "children": {
          "$$typeof": Symbol(react.transitional.element),
          "_owner": null,
          "_store": {},
          "key": null,
          "props": {
            "children": {
              "$$typeof": Symbol(react.transitional.element),
              "_owner": null,
              "_store": {},
              "key": null,
              "props": {
                "children": {
                  "$$typeof": Symbol(react.transitional.element),
                  "_owner": null,
                  "_store": {},
                  "key": null,
                  "props": {
                    "children": "Test Child",
                  },
                  "type": "div",
                },
                "headers": Headers {},
              },
              "type": [Function],
            },
            "session": null,
          },
          "type": [Function],
        },
        "className": "font-sans --font-sans min-h-screen",
      },
      "type": "body",
    },
    "lang": "en",
  },
  "type": "html",
}
`;

exports[`RootLayout > renders the layout with session provider and children 1`] = `
{
  "$$typeof": Symbol(react.transitional.element),
  "_owner": null,
  "_store": {},
  "key": null,
  "props": {
    "children": {
      "$$typeof": Symbol(react.transitional.element),
      "_owner": null,
      "_store": {},
      "key": null,
      "props": {
        "children": {
          "$$typeof": Symbol(react.transitional.element),
          "_owner": null,
          "_store": {},
          "key": null,
          "props": {
            "children": {
              "$$typeof": Symbol(react.transitional.element),
              "_owner": null,
              "_store": {},
              "key": null,
              "props": {
                "children": {
                  "$$typeof": Symbol(react.transitional.element),
                  "_owner": null,
                  "_store": {},
                  "key": null,
                  "props": {
                    "children": "Test Child",
                  },
                  "type": "div",
                },
                "headers": Headers {},
              },
              "type": [Function],
            },
            "session": {
              "expires": "2099-01-01T00:00:00.000Z",
              "user": {
                "email": "test@example.com",
                "id": "test-user-id",
              },
            },
          },
          "type": [Function],
        },
        "className": "font-sans --font-sans min-h-screen",
      },
      "type": "body",
    },
    "lang": "en",
  },
  "type": "html",
}
`;
